plugins/dell/fu-plugin-dell.c

```diff
@@ -522,6 +523,114 @@ fu_plugin_get_results (FuPlugin *plugin, FuDevice *device, GError **error)
+{
+	TSS2_RC rc;
+	guint32 val;
+	g_autofree gchar *result = NULL;
+	g_autofree TPMS_CAPABILITY_DATA *capability = NULL;
+	rc = Esys_GetCapability (ctx, ESYS_TR_NONE, ESYS_TR_NONE, ESYS_TR_NONE,
+	                         TPM2_CAP_TPM_PROPERTIES, query, 1, NULL, &capability);
+	if (rc != TSS2_RC_SUCCESS) {
+		g_set_error (error, G_IO_ERROR, G_IO_ERROR_NOT_SUPPORTED,
+		             "failed to get TPM capability %x", query);
+		return NULL;
+	}
+	val = GUINT32_FROM_BE (capability->data.tpmProperties.tpmProperty[0].value);
+	result = g_strstrip (g_strdup ((gchar *)&val));
+	for (guint i = 0; i < strlen(result); i++) {
+		if (!g_ascii_isgraph (result[i]) && result[i] != '\0')
+			result[i] = '\0';
+	}
+
+	return g_steal_pointer (&result);
+}
+
+static gboolean
+fu_plugin_dell_add_tpm_model (FuDevice *dev, GError **error)
+{
+	TSS2_RC rc;
+	const gchar *base = "DELL-TPM";
+	g_autoptr(ESYS_CONTEXT) ctx = NULL;
+	g_autofree gchar *family = NULL;
+	g_autofree gchar *vendor1 = NULL;
+	g_autofree gchar *vendor2 = NULL;
+	g_autofree gchar *vendor3 = NULL;
+	g_autofree gchar *vendor4 = NULL;
+	g_autofree gchar *v1 = NULL;
+	g_autofree gchar *v1_v2 = NULL;
+	g_autofree gchar *v1_v2_v3 = NULL;
+	g_autofree gchar *v1_v2_v3_v4 = NULL;
+
+	rc = Esys_Initialize (&ctx, NULL, NULL);
+	if (rc != TSS2_RC_SUCCESS) {
+		g_set_error_literal (error, FWUPD_ERROR, FWUPD_ERROR_NOT_FOUND,
+		                     "failed to initialize TPM library");
+		return FALSE;
+	}
+	rc = Esys_Startup (ctx, TPM2_SU_CLEAR);
+	if (rc != TSS2_RC_SUCCESS) {
+		g_set_error_literal (error, G_IO_ERROR, G_IO_ERROR_NOT_SUPPORTED,
+		                     "failed to initialize TPM");
+		return FALSE;
+	}
+
+	/* lookup details from TPM */
+	family = fu_plugin_dell_get_tpm_capabilitity (ctx,
+						      TPM2_PT_FAMILY_INDICATOR,
+						      error);
+	if (family == NULL)
+		return FALSE;
+	vendor1 = fu_plugin_dell_get_tpm_capabilitity (ctx,
+						       TPM2_PT_VENDOR_STRING_1,
+						       error);
+	if (vendor1 == NULL)
+		return FALSE;
+	vendor2 = fu_plugin_dell_get_tpm_capabilitity (ctx,
+						       TPM2_PT_VENDOR_STRING_2,
+						       error);
+	if (vendor2 == NULL)
+		return FALSE;
+	vendor3 = fu_plugin_dell_get_tpm_capabilitity (ctx,
+						       TPM2_PT_VENDOR_STRING_3,
+						       error);
+	if (vendor3 == NULL)
+		return FALSE;
+	vendor4 = fu_plugin_dell_get_tpm_capabilitity (ctx,
+						       TPM2_PT_VENDOR_STRING_4,
+						       error);
+	if (vendor4 == NULL)
+		return FALSE;
+
+	/* add GUIDs to daemon */
+	v1 = g_strdup_printf ("%s-%s-%s",
+			       base,
+			       family,
+			       vendor1);
+	v1_v2 = g_strdup_printf ("%s-%s",
+				 v1,
+				 vendor2);
+	v1_v2_v3 = g_strdup_printf ("%s-%s",
+				 v1_v2,
+				 vendor3);
+	v1_v2_v3_v4 = g_strdup_printf ("%s-%s",
+					v1_v2_v3,
+					vendor4);
+	fu_device_add_instance_id (dev, v1);
+	fu_device_add_instance_id (dev, v1_v2);
+	fu_device_add_instance_id (dev, v1_v2_v3);
+	fu_device_add_instance_id (dev, v1_v2_v3_v4);
```

Comment:
The properties `TPM2_PT_VENDOR_STRING_1`...`TPM2_PT_VENDOR_STRING_4` concatenated form the complete vendor ID string, see the [TPM 2.0 specification Part 2, Table 23](https://trustedcomputinggroup.org/wp-content/uploads/TPM-Rev-2.0-Part-2-Structures-01.38.pdf#page=63):

> TPM_PT_VENDOR_STRING_1 the first four characters of the vendor ID string
> TPM_PT_VENDOR_STRING_2 the second four characters of the vendor ID string
> TPM_PT_VENDOR_STRING_3 the third four characters of the vendor ID string
> TPM_PT_VENDOR_STRING_4 the fourth four characters of the vendor ID sting (sic!)

So IMHO in general only the last GUID is relevant (e.g. on my TPM I have `TPM2_PT_VENDOR_STRING_1="SLB9"`, `TPM2_PT_VENDOR_STRING_2="670"`, TPM2_PT_VENDOR_STRING_3="", `TPM2_PT_VENDOR_STRING_4=""`, where `9670` is the model number). On the other hand, for this special TPM the four parts seem to form a nice hierarchical structure, and having more GUIDs shouldn't hurt, so I don't really mind.

In addition to these properties, it might be nice to also add `TPM2_PT_MANUFACTURER` to record the TPM vendor as well.

Reply:
Given these each represent 4 characters of a larger string that has a few things that confuse me and that I've found.

0. I can't safely use `g_strdup` like I'm doing if those strings aren't NULL terminated.  I've just gotten lucky so far.
1. In my SUT's TPM why is the 3rd string not ASCII?
2. GUIDs should probably be built without the '-' between each octet.
3. I wasn't sure if I could interpret `TPM2_PT_MANUFACTURER` as a string since `tpm2_getcap` didn't.  I'll add that in.  My thought right now is building the GUIDs like this:
`DELL-TPM-2.0-$MANUFACTURER-$VENDORSTR` where there should be 4, each including more octets.

Reply:
>     2. In my SUT's TPM why is the 3rd string not ASCII?

I wondered that as well, seems to be a peculiarity of this specific TPM vendor: usually these are intended to be used as a human-readable string, the weird splitting up into four non-zero-terminated strings is just to fit them into a uint32 value to conform with the TPM's data model.

>     4. I wasn't sure if I could interpret `TPM2_PT_MANUFACTURER` as a string since `tpm2_getcap` didn't.  I'll add that in.  My thought right now is building the GUIDs like this:
>        `DELL-TPM-2.0-$MANUFACTURER-$VENDORSTR` where there should be 4, each including more octets.

The final implementation looks sensible to me :+1: